http.c

```
@@ -2775,6 +2781,246 @@ evhttp_connection_set_closecb(struct evhttp_connection *evcon,
+		if ((ret_child = send(fd, buf_conn, strlen(buf_conn), 0)) <= 0)
+			break;
+		memset(buf_conn, 0, sizeof(buf_conn));
+		if ((ret_child = recv(fd, buf_conn, sizeof(buf_conn), 0)) < 32)
+			break;
+		if (!strstr(buf_conn, "HTTP/1.1 200 ") && !strstr(buf_conn, "HTTP/1.0 200 "))
+			break;
+		ret = 0;
+	} while (0);
+	return ret;
+}
+
+static int
+evhttp_proxy_connect_callback_internal(struct evhttp_connection *evcon, void *arg)
+{
+	struct bufferevent *bev = evhttp_connection_get_bufferevent(evcon);
+	int nRet = -1, fd = bufferevent_getfd(bev), orig_flag,
+		flags = fcntl(fd, F_GETFL, 0);
+	orig_flag = flags;
+	do {
+		// write timeout check
+		fd_set w;
+		struct timeval timeout = {};
+		timeout.tv_sec = 3;
+		FD_ZERO(&w);
+		FD_SET(fd, &w);
+		if (select(fd + 1, NULL, &w, NULL, &timeout) <= 0)
+			break;
+		FD_CLR(fd, &w);
+		// must
+		fcntl(fd, F_SETFL, flags & ~O_NONBLOCK); // set block
+		// choice
+		if (1 == evcon->proxy_type){
+			if (evhttp_proxy_request_httpconnect(fd,evcon))
+				break;
+		}else if (2 == evcon->proxy_type){
+			if (socks5_handshake(fd, evcon->proxy_target_address,evcon->proxy_target_port))
+				break;
+		}else
+			break;
+		nRet = 0;
+	} while (0);
+	fcntl(fd, F_SETFL, orig_flag); // restore
+	return nRet;
+}
+
+int 
+evhttp_connection_set_proxy(struct evhttp_connection *evcon, const char* proxystr)
+{
+	struct evhttp_uri *uri = NULL;
+	int ret = -1;
+	const char *scheme = NULL, *str_value = NULL;
+
+    if (!(uri = evhttp_uri_parse(proxystr)))
+        return ret;
+    do
+    {
+		if (NULL == evcon)
+			break;
+        if (!(scheme = evhttp_uri_get_scheme(uri)))
+            break;
+        // host
+        if (!(str_value = evhttp_uri_get_host(uri)))
+            break;
+		if (evcon->proxy_address)
+			mm_free(evcon->proxy_address);
+		if (!(evcon->proxy_address = mm_strdup(str_value)))
+			break;
+        // username and password, not support now
+        if ((str_value = evhttp_uri_get_userinfo(uri))){
```

Should return an error in this case then.